# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: widget/gtk/WaylandBuffer.cpp
# Commit: 3927f21c92c9
# Full Hash: 3927f21c92c941c9efd0b00b0dc4cfc2b905452f
# Author: stransky <stransky@redhat.com>
# Date: 2025-09-08 08:22:14
# Description:
#   Bug 1979106 [Wayland] Don't recycle buffer transactions across WaylandSurfaces r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D263545
# ==============================================================================

diff -r 2a24e5601893 -r 3927f21c92c9 widget/gtk/WaylandBuffer.cpp
--- a/widget/gtk/WaylandBuffer.cpp	Mon Sep 08 02:52:34 2025 +0000
+++ b/widget/gtk/WaylandBuffer.cpp	Mon Sep 08 04:20:40 2025 +0000
@@ -99,9 +99,10 @@
   return false;
 }
 
-BufferTransaction* WaylandBuffer::GetTransaction() {
+BufferTransaction* WaylandBuffer::GetTransaction(
+    const WaylandSurfaceLock& aSurfaceLock) {
   for (const auto& transaction : mBufferTransactions) {
-    if (transaction->IsDetached()) {
+    if (transaction->CanRecycle(aSurfaceLock.GetWaylandSurface())) {
       LOGWAYLAND("WaylandBuffer::GetTransaction() [%p] reuse transaction [%d]",
                  (void*)this, (int)mBufferTransactions.Length());
       return transaction;
@@ -120,9 +121,9 @@
 
   LOGWAYLAND(
       "WaylandBuffer::GetTransaction() create new [%p] wl_buffer [%p] "
-      "transactions [%d] external buffer [%d]",
+      "transactions [%d] external buffer [%d] WaylandSurface [%p]",
       (void*)this, buffer, (int)mBufferTransactions.Length(),
-      !!mExternalWlBuffer);
+      !!mExternalWlBuffer, aSurfaceLock.GetWaylandSurface());
 
   auto* transaction = new BufferTransaction(this, buffer, !!mExternalWlBuffer);
   mBufferTransactions.AppendElement(transaction);
@@ -332,12 +333,15 @@
     const WaylandSurfaceLock& aSurfaceLock) {
   LOGWAYLAND(
       "BufferTransaction::BufferBorrow() [%p] widget [%p] WaylandSurface [%p] "
-      "(old %p) "
       "WaylandBuffer [%p]",
       this, aSurfaceLock.GetWaylandSurface()->GetLoggingWidget(),
-      aSurfaceLock.GetWaylandSurface(), mSurface.get(), mBuffer.get());
+      aSurfaceLock.GetWaylandSurface(), mBuffer.get());
 
+  MOZ_DIAGNOSTIC_ASSERT(
+      !mSurface || mSurface == aSurfaceLock.GetWaylandSurface(),
+      "Can't transfer transaction between WaylandSurfaces!");
   MOZ_DIAGNOSTIC_ASSERT(mBufferState == BufferState::Detached);
+
   mSurface = aSurfaceLock.GetWaylandSurface();
 
   // We don't take reference to this. Some compositors doesn't send